cpp/src/spatial/hausdorff.cu

```diff
@@ -0,0 +1,200 @@
+
+        T min_dist_sqrd = 1e20;
+
+        int num_points_in_b = space_b_end - space_b_begin;
+
+        if (threadIdx.x < num_points_in_b)
+        {
+            T point_b_x = xs[space_b_begin + threadIdx.x];
+            T point_b_y = ys[space_b_begin + threadIdx.x];
+
+            for (int i = space_a_begin; i < space_a_end; i++)
+            {
+                T point_a_x = xs[i];
+                T point_a_y = ys[i];
+                T dist_sqrd = magnitude_squared(point_b_x - point_a_x, point_b_y - point_a_y);
+
+                min_dist_sqrd = min(min_dist_sqrd, dist_sqrd);
+            }
+        }
+
+        if (min_dist_sqrd > 1e10)
+        {
+            min_dist_sqrd = -1;
+        }
+
+        __shared__ T dist_sqrd[MAX_NUM_THREADS_PER_BLOCK];
+
+        dist_sqrd[threadIdx.x] = threadIdx.x < num_points_in_b ? min_dist_sqrd : -1;
+
+        for (int offset = blockDim.x / 2; offset > 0; offset >>= 1)
+        {
+            __syncthreads();
+
+            if (threadIdx.x < offset)
+            {
+                dist_sqrd[threadIdx.x] = max(dist_sqrd[threadIdx.x],
+                                                dist_sqrd[threadIdx.x + offset]);
+            }
+        }
+
+        if (threadIdx.x == 0)
+        {
+            results[block_idx] = (dist_sqrd[0] < 0) ? 1e10 : sqrt(dist_sqrd[0]);
+        }
+    }
+}
+
+struct hausdorff_functor
+{
+    template<typename T, typename... Args>
+    std::enable_if_t<not std::is_floating_point<T>::value, std::unique_ptr<cudf::column>>
+    operator()(Args&&...)
+    {
+        CUSPATIAL_FAIL("Non-floating point operation is not supported");
+    }
+
+    template<typename T>
+    std::enable_if_t<std::is_floating_point<T>::value, std::unique_ptr<cudf::column>>
+    operator()(cudf::column_view const& xs,
+               cudf::column_view const& ys,
+               cudf::column_view const& points_per_space,
+               rmm::mr::device_memory_resource *mr,
+               cudaStream_t stream)
+    {
+        auto tid = cudf::experimental::type_to_id<T>();
+        auto result = cudf::make_fixed_width_column(cudf::data_type{ tid },
+                                                    points_per_space.size() * points_per_space.size(),
+                                                    cudf::mask_state::UNALLOCATED,
+                                                    stream,
+                                                    mr);
+
+        if (result->size() == 0)
+        {
+            return result;
+        }
+
+        auto space_offsets = rmm::device_vector<cudf::size_type>(points_per_space.size());
+
+        thrust::inclusive_scan(rmm::exec_policy(stream)->on(stream),
+                               points_per_space.begin<cudf::size_type>(),
+                               points_per_space.end<cudf::size_type>(),
+                               space_offsets.begin());
+
+        CUSPATIAL_EXPECTS(space_offsets.back() <= xs.size(),
```

Comment:
I'm happy to see input validation, but I think Mark won't like the `.back()` here since we determined that this relies on device synchronization. I'm not sure how to get around this other than using `cudaMemcpyAsync()` and `cudaStreamSync()` directly though.

Reply:
Ah, yes, I wrote this at 3am. It looked like a good idea at the time. :laughing:

Reply:
This is not the only use of back() in RAPIDS...  Can't fix it nicely until device_uvector is done.